### ISSUE TITLE
Kraken FIAT funding 

### ISSUE THREAD

**Issue by npomfret** [ccxt/ccxt#4642]
@kroitor The only fiat currencies that Kraken allows deposits for at the moment are Euros, and the only FIAT currencies allowed for Euros (not even Canadian Dollars are enabled).

Would you accept a PR with these values hard coded?

**Comment by kroitor** [comment]
Hi!

Can you provide some formal info on that? 

Namely, how do you determine this: 

> The only fiat currencies that Kraken allows deposits for at the moment are Euros

If they have a list of currently active currencies somewhere, we'd rather parse it, instead of hardcoding. In general we deprecated hardcoding, because if Kraken enables those at some point later – we might miss that, unless we do the proper automation to detect for that.

**Comment by npomfret** [comment]
> we'd rather parse it, instead of hardcoding

Understood, but I can't find a list anywhere.  They have a support article that says you need to log into  to see the current valid withdrawal methods.  Which I did, and euros is the only one available right now. 

I know hard coding isn't ideal, but in this case I believe its the best option.

**Comment by npomfret** [comment]
@kroitor appreciate its controversial and certainly not ideal, but i've added a PR - https://github.com/ccxt/ccxt/pull/4645 - understand if you can't accept.

**Comment by npomfret** [comment]
@kroitor given that we represent percentages as decimals, how do you tell if a fee is a fixed fee or a percentage?  Lots of withdrawal fees for fiat currencies are percentages.

**Comment by kroitor** [comment]
@npomfret we have these flags in the `fee` structure that goes into currencies upon parsing: https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L163

**Comment by npomfret** [comment]
In the example you show they are all fixed and make as so with `'percentage': false`.  But bitfinex for example has both fixed and percentage based withdrawal fees.  See _bank wire_ here: (https://www.bitfinex.com/fees).

**Comment by kroitor** [comment]
> But bitfinex for example has both fixed and percentage based withdrawal fees.

Yep, and for those currencies we need to set it to `true` in fetchCurrencies upon parsing, if the exchange allows to determine that. However, because not all exchanges provide the necessary info, and also because this is still a work in progress, some of those flags may be inconsistent here and there. If you help us find some ways to fix some of those inconsistencies – that would be really great. Ways to detect that properly, i mean, not to hardcode that, of course ))

**Comment by npomfret** [comment]
Will do.  I think with kraken though, hard coding is the only way for now.

**Comment by npomfret** [comment]
Also, do you have an _ideal_ model for currency funding fees?  If so, what should it look like?  I will help to start making them consistent.

**Comment by kroitor** [comment]
> Also, do you have an _ideal_ model for currency funding fees?

TBH, we are still in the process of designing it – currently, the structure has a plain flat `fee` field with withdrawal fee rate for each currency... However, some exchanges have deposit fees, especially with fiat currencies. Some other exchanges have a min flat fixed fee and a percentage on top, like "0.1%, min $1", depending on the amount (or similar to that). Others might have a mandatory flat fee + a percentage you pay "$1 + 0.1% of the amount" regardless of the amount. So, it isn't very easy to devise a universally good, flexible and convenient-to-use structure. Your thoughts are welcome! )

**Comment by npomfret** [comment]
My thoughts, for what they're worth:

Don't bother trying to make a perfect model.  I've seen an exchange that charges fees in a currency that isn't even the currency you're dealing with.  It's close to impossible to come up with a universally _correct_ model.  The important bits as I see them are:

Ability to easily specify fixed v's percentage fees. I'd go for `"1%"` over `0.01`, then you can use numeric values to specify fixed fees (without needing extra markup to distinguish): `BTC: 0.01, EUR: "1%"...`.  The assumption of course is that the fee is charge in the same currency.

Ability to provide options: `EUR: {'bank wire': "1%", 'credit card': "3%"}, ...`

Ability to provide more _complex_ fees: `EUR: {'bank wire': {rate: "1%", fixed: 5]}, ...`

I imagine that would cover the vast majority of cases.

**Comment by npomfret** [comment]
Anyway... https://github.com/ccxt/ccxt/pull/4653

**Comment by npomfret** [comment]
@kroitor regarding the question about representing mixed fees (fixed / percentage).  `therock` has _percentage_ withdrawal fees on just two currencies (XRP and FDZ) (see [here](https://www.therocktrading.com/en/pages/fees)).  All others are fixed.

I don't think there's an api for this data.  Currently its hard coded, except the 2 in question:

```
 'fees': {
                'trading': {
                    'maker': 0.2 / 100,
                    'taker': 0.2 / 100,
                },
                'funding': {
                    'tierBased': false,
                    'percentage': false,
                    'withdraw': {
                        'BTC': 0.0005,
                        'BCH': 0.0005,
                        'PPC': 0.02,
                        'ETH': 0.001,
                        'ZEC': 0.001,
                        'LTC': 0.002,
                        'EUR': 2.5,  // worst-case scenario: https://therocktrading.com/en/pages/fees
                    },
                    'deposit': {
                        'BTC': 0,
                        'BCH': 0,
                        'PPC': 0,
                        'ETH': 0,
                        'ZEC': 0,
                        'LTC': 0,
                        'EUR': 0,
                    },
                },
            },
```

Do you have a suggestion as to a _correct_ way to go about implementing this?

**Comment by kroitor** [comment]
> Do you have a suggestion as to a _correct_ way to go about implementing this?

To be honest, not yet, we are in the process of working it out.

**Comment by kroitor** [comment]
It gets more complicated if we also account for this as well: https://github.com/ccxt/ccxt/pull/1814#issuecomment-363827160

**Comment by npomfret** [comment]
Yeah, it's a tricky one alright.  Whatever the solution I think a 'fee' needs to be modelled as an object, rather than a number.  Something like

    'funding': {
                    'withdraw': {
                        'BTC': { 'fixed': 0.0005 },
                        'XRP': { 'rate': '0.05%' },
                        'FDZ': { 'rate': 0.0001 },
                        'EUR': { 'rate': '1%', fixed: 5, min: 1000 },
                        'USD': { 'rate': 0, fixed: 80 }
                        ...

So forget the _global_ 'percentage' flag and embed that knowledge inside the fee object. I like being able to express a percentage as a string (eg '1%"), but with this approach both strings and decimals can be supported.

I guess fee's for trading are a little different because the currency that the fee is charged in cannot be implied (as it can be with funding).
